src/GraphQL/Http/DocumentWriter.cs

```diff
@@ -26,13 +34,62 @@ public DocumentWriter(bool indent)
 
         public DocumentWriter(Formatting formatting, JsonSerializerSettings settings)
         {
-            _formatting = formatting;
-            _settings = settings;
+            _serializer = JsonSerializer.CreateDefault(settings);
+            _serializer.Formatting = formatting;
+        }
+
+        public Task WriteAsync(Stream stream, ExecutionResult value)
+        {
+            using (var writer = new StreamWriter(stream, Utf8Encoding, 1024, true))
+            using (var jsonWriter = new JsonTextWriter(writer))
+            {
+                _serializer.Serialize(jsonWriter, value);
+            }
+            return TaskExtensions.CompletedTask;
         }
 
         public string Write(object value)
         {
-            return JsonConvert.SerializeObject(value, _formatting, _settings);
+            return this.WriteToStringAsync((ExecutionResult)value).GetAwaiter().GetResult();
+        }
+    }
+
+    public static class DocumentWriterExtensions
+    {
+        /// <summary>
+        /// Writes the <paramref name="value"/> to string.
+        /// </summary>
+        /// <param name="writer"></param>
+        /// <param name="value"></param>
+        /// <param name="encoding"></param>
+        /// <returns></returns>
+        public static async Task<string> WriteToStringAsync(this IDocumentWriter writer,
+            ExecutionResult value,
+            Encoding encoding = null)
+        {
+            var resolvedEncoding = encoding ?? Encoding.UTF8;
+            using (var stream = new MemoryStream())
+            {
+                await writer.WriteAsync(stream, value).ConfigureAwait(false);
+#if NET45
+                var length = (int)stream.Length;
+                var offset = (int)stream.Seek(0, SeekOrigin.Begin);
+                var buffer = stream.GetBuffer();
+
+                return resolvedEncoding.GetString(buffer, offset, length - offset);
+#else
+                // Will succeed since we use default MemoryStream constructor
+                stream.TryGetBuffer(out var buffer);
+                return resolvedEncoding.GetString(buffer.Array, buffer.Offset, buffer.Count);
+#endif
+            }
+        }
+
+        public static async Task<JObject> WriteToJObjectAsync(this IDocumentWriter writer,
+            ExecutionResult value)
+        {
+            var json = await writer.WriteToStringAsync(value);
+            return JObject.Parse(json);
```

Comment:
Did you try using `JObject.FromObject`?  Does it serialize correctly?  It may not, though is worth a shot.

```
public static JObject FromObject(
	Object o,
	JsonSerializer jsonSerializer
)
```

Reply:
It does work. 
The reason I chose not to call JObject.Parse is that
internally, JObject.FromObject just serializes into JTokenWriter anyway. 
https://github.com/JamesNK/Newtonsoft.Json/blob/master/Src/Newtonsoft.Json/Linq/JToken.cs#L1881.
Since DocumentWriter might potentially customize the serialization workflow, I chose to use this approach. In addition, there is no need to pass JsonSerializer here - the one in DocumentWriter will be reused.

Reply:
I mentioned the one with the serializer to use if we were avoiding calling the `WriteToStringAsync`.  It would seem like calling `FromObject` would be much faster than serializing to string then re-parsing.

Reply:
``` ini

BenchmarkDotNet=v0.11.0, OS=Windows 10.0.17134.254 (1803/April2018Update/Redstone4)
Intel Core i7-7700HQ CPU 2.80GHz (Max: 0.80GHz) (Kaby Lake), 1 CPU, 8 logical and 4 physical cores
.NET Core SDK=2.1.401
  [Host]     : .NET Core 2.1.2 (CoreCLR 4.6.26628.05, CoreFX 4.6.26629.01), 64bit RyuJIT
  DefaultJob : .NET Core 2.1.2 (CoreCLR 4.6.26628.05, CoreFX 4.6.26629.01), 64bit RyuJIT


```
|              Method |     Mean |    Error |   StdDev |     Gen 0 |     Gen 1 |    Gen 2 | Allocated |
|-------------------- |---------:|---------:|---------:|----------:|----------:|---------:|----------:|
|          FromObject | 122.1 ms | 2.428 ms | 5.226 ms | 4800.0000 | 2000.0000 | 600.0000 |  25.81 MB |
| UsingDocumentWriter | 131.9 ms | 2.634 ms | 4.882 ms | 6000.0000 | 2500.0000 | 500.0000 |   36.3 MB |

FromObject will be faster. Both options are very slow in general and allocate a lot of memory (Compare this with the results of direct stream writing, it's using the same test set). I think we need to consider a better alternative than passing JObjects for WebSocket API.

Reply:
This might very well be the case. Then maybe the `WriteToJObjectAsync` method shouldn't be included in order to not encourage it? Sorry for leading you out on a wild goose chase here. 🍕

Reply:
Yeah, I think we need to come up with something else for websockets. I'll remove the method for now.